src/systemctl/systemctl-list-unit-files.c

```diff
@@ -11,25 +12,37 @@
 
 static int compare_unit_file_list(const UnitFileList *a, const UnitFileList *b) {
         const char *d1, *d2;
+        _cleanup_free_ char *a_id = NULL, *b_id = NULL;
+        int r;
 
         d1 = strrchr(a->path, '.');
         d2 = strrchr(b->path, '.');
 
         if (d1 && d2) {
-                int r;
-
                 r = strcasecmp(d1, d2);
                 if (r != 0)
                         return r;
         }
 
-        return strcasecmp(basename(a->path), basename(b->path));
+        r = path_extract_filename(a->path, &a_id);
+        if (r < 0)
+                return r;
+
+        r = path_extract_filename(b->path, &b_id);
+        if (r < 0)
+                return r;
+
+        return strcasecmp(a_id, b_id);
 }
 
 static bool output_show_unit_file(const UnitFileList *u, char **states, char **patterns) {
+        _cleanup_free_ char *unit_filename = NULL;
         assert(u);
 
-        if (!strv_fnmatch_or_empty(patterns, basename(u->path), FNM_NOESCAPE))
+        if (path_extract_filename(u->path, &unit_filename) < 0)
+                return false;
```

Comment:
hm, please change output_show_unit_file() to properly return errors, i.e. < 0 for error, == 0 for false, and > 0 for true.

and then make sure the callers know how to handle this.

Reply:
I did this in a different commit.  I wasn't sure if I should replace more of the `false`s with actual errors.  For instance, the `strv_fnmatch_or_empty` branch could use `-ENOENT`.  But since it isn't really an "error" IMO but rather a lack of match, I didn't change that.